File: csharp/NeoC.Core/Parser.cs
namespace NeoC.Core;

// Parser recursivo descendente con estado en ref struct.
public ref struct Parser
{
    private readonly IReadOnlyList<Token> _tokens;
    private int _i;

    public Parser(IReadOnlyList<Token> tokens)
    {
        _tokens = tokens;
        _i = 0;
    }

    private Token Cur => _tokens[_i];

    public ProgramNode ParseProgram()
    {
        var imports = new List<ImportNode>();
        var aliases = new List<AliasNode>();
        var items = new List<ItemNode>();

        while (Cur.Kind != TokenKind.Eof)
        {
            if (Match(TokenKind.Import)) imports.Add(ParseImport());
            else if (Match(TokenKind.Alias)) aliases.Add(ParseAlias());
            else items.Add(ParseItem());
        }

        return new(imports, aliases, items);
    }

    private ImportNode ParseImport()
    {
        var path = Expect(TokenKind.StringLit, "string en import").Lexeme;
        Expect(TokenKind.Semicolon, "';' en import");
        return new(path);
    }

    private AliasNode ParseAlias()
    {
        var name = Expect(TokenKind.Ident, "identificador de alias").Lexeme;
        Expect(TokenKind.Equal, "'=' en alias");
        var path = Expect(TokenKind.StringLit, "string en alias").Lexeme;
        Expect(TokenKind.Semicolon, "';' en alias");
        return new(name, path);
    }

    private ItemNode ParseItem()
    {
        var isExtern = Match(TokenKind.Extern);
        if (Match(TokenKind.Struct)) return ItemNode.FromStruct(ParseStruct(isExtern));
        if (Match(TokenKind.Fn)) return ItemNode.FromFunction(ParseFunction(isExtern));
        throw Error("item inválido");
    }

    private StructNode ParseStruct(bool isExtern)
    {
        var name = Expect(TokenKind.Ident, "nombre de struct").Lexeme;
        Expect(TokenKind.LBrace, "'{' en struct");

        var fields = new List<FieldNode>();
        while (Cur.Kind != TokenKind.RBrace)
        {
            var ty = ParseType();
            var fieldName = Expect(TokenKind.Ident, "campo").Lexeme;
            Expect(TokenKind.Semicolon, "';' en campo");
            fields.Add(new(ty, fieldName));
        }

        Expect(TokenKind.RBrace, "'}' en struct");
        Expect(TokenKind.Semicolon, "';' tras struct");
        return new(name, fields, isExtern);
    }

    private FunctionNode ParseFunction(bool isExtern)
    {
        var name = Expect(TokenKind.Ident, "nombre de función").Lexeme;
        Expect(TokenKind.LParen, "'('");

        var parameters = new List<ParamNode>();
        if (Cur.Kind != TokenKind.RParen)
        {
            do
            {
                var ty = ParseType();
                var pName = Expect(TokenKind.Ident, "parámetro").Lexeme;
                parameters.Add(new(ty, pName));
            } while (Match(TokenKind.Comma));
        }

        Expect(TokenKind.RParen, "')'");
        Expect(TokenKind.Arrow, "'->'");
        var ret = ParseType();

        if (Match(TokenKind.Semicolon))
        {
            return new(name, parameters, ret, null, true);
        }

        Expect(TokenKind.LBrace, "'{' en cuerpo");
        var body = new List<StmtNode>();
        while (Cur.Kind != TokenKind.RBrace)
        {
            body.Add(ParseStmt());
        }
        Expect(TokenKind.RBrace, "'}' en cuerpo");

        return new(name, parameters, ret, body, isExtern);
    }

    private StmtNode ParseStmt()
    {
        if (Match(TokenKind.Let))
        {
            var ty = ParseType();
            var name = Expect(TokenKind.Ident, "variable").Lexeme;
            Expect(TokenKind.Equal, "'=' en let");
            var expr = ParseExpr();
            Expect(TokenKind.Semicolon, "';' en let");
            return StmtNode.FromLet(new(ty, name, expr));
        }

        if (Match(TokenKind.Return))
        {
            var expr = ParseExpr();
            Expect(TokenKind.Semicolon, "';' en return");
            return StmtNode.FromReturn(new(expr));
        }

        var value = ParseExpr();
        Expect(TokenKind.Semicolon, "';' en expresión");
        return StmtNode.FromExpr(new(value));
    }

    private ExprNode ParseExpr() => ParseAddSub();

    private ExprNode ParseAddSub()
    {
        var node = ParseMulDiv();
        while (Cur.Kind is TokenKind.Plus or TokenKind.Minus)
        {
            var op = Cur.Kind == TokenKind.Plus ? BinaryOp.Add : BinaryOp.Sub;
            Next();
            node = ExprNode.FromBinary(new(op, node, ParseMulDiv()));
        }

        return node;
    }

    private ExprNode ParseMulDiv()
    {
        var node = ParsePrimary();
        while (Cur.Kind is TokenKind.Star or TokenKind.Slash)
        {
            var op = Cur.Kind == TokenKind.Star ? BinaryOp.Mul : BinaryOp.Div;
            Next();
            node = ExprNode.FromBinary(new(op, node, ParsePrimary()));
        }
        return node;
    }

    private ExprNode ParsePrimary()
    {
        if (Cur.Kind == TokenKind.IntLit)
        {
            var value = long.Parse(Cur.Lexeme);
            Next();
            return ExprNode.FromInt(new(value));
        }

        if (Cur.Kind == TokenKind.Ident)
        {
            var name = Cur.Lexeme;
            Next();
            if (Match(TokenKind.LParen))
            {
                var args = new List<ExprNode>();
                if (Cur.Kind != TokenKind.RParen)
                {
                    do { args.Add(ParseExpr()); } while (Match(TokenKind.Comma));
                }
                Expect(TokenKind.RParen, "')' en llamada");
                return ExprNode.FromCall(new(name, args));
            }

            return ExprNode.FromVar(new(name));
        }

        if (Match(TokenKind.LParen))
        {
            var inner = ParseExpr();
            Expect(TokenKind.RParen, "')' de agrupación");
            return inner;
        }

        throw Error("expresión inválida");
    }

    private TypeRef ParseType()
    {
        var id = Expect(TokenKind.Ident, "tipo").Lexeme;
        return id switch
        {
            "int" => TypeRef.Int,
            "void" => TypeRef.Void,
            _ => TypeRef.Named(id)
        };
    }

    private bool Match(TokenKind kind)
    {
        if (Cur.Kind == kind)
        {
            Next();
            return true;
        }
        return false;
    }

    private Token Expect(TokenKind kind, string msg)
        => Cur.Kind == kind ? Next() : throw Error($"Se esperaba {msg}");

    private Token Next()
    {
        var t = Cur;
        if (_i < _tokens.Count - 1) _i++;
        return t;
    }

    private Exception Error(string msg) => new($"{msg} en {Cur.Line}:{Cur.Column}");
}

File: src/parser.rs
use crate::ast::*;
use crate::token::{Token, TokenKind};

pub struct Parser {
    tokens: Vec<Token>,
    pos: usize,
}

impl Parser {
    pub fn new(tokens: Vec<Token>) -> Self {
        Self { tokens, pos: 0 }
    }

    fn cur(&self) -> &Token {
        &self.tokens[self.pos]
    }

    fn bump(&mut self) {
        if self.pos < self.tokens.len() - 1 {
            self.pos += 1;
        }
    }

    fn eat(&mut self, kind: &TokenKind) -> bool {
        if &self.cur().kind == kind {
            self.bump();
            true
        } else {
            false
        }
    }

    fn expect(&mut self, kind: &TokenKind, msg: &str) -> Result<(), String> {
        if self.eat(kind) {
            Ok(())
        } else {
            Err(format!("{} en {}:{}", msg, self.cur().line, self.cur().col))
        }
    }

    fn expect_ident(&mut self, what: &str) -> Result<String, String> {
        match &self.cur().kind {
            TokenKind::Ident(s) => {
                let out = s.clone();
                self.bump();
                Ok(out)
            }
            _ => Err(format!(
                "Se esperaba {} en {}:{}",
                what,
                self.cur().line,
                self.cur().col
            )),
        }
    }

    fn parse_type(&mut self) -> Result<TypeName, String> {
        match &self.cur().kind {
            TokenKind::Ident(s) if s == "int" => {
                self.bump();
                Ok(TypeName::Int)
            }
            TokenKind::Ident(s) if s == "void" => {
                self.bump();
                Ok(TypeName::Void)
            }
            TokenKind::Ident(s) => {
                let n = s.clone();
                self.bump();
                Ok(TypeName::Named(n))
            }
            _ => Err(format!(
                "Tipo inválido en {}:{}",
                self.cur().line,
                self.cur().col
            )),
        }
    }

    pub fn parse_program(&mut self) -> Result<Program, String> {
        let mut imports = Vec::new();
        let mut aliases = Vec::new();
        let mut items = Vec::new();

        while self.cur().kind != TokenKind::Eof {
            match self.cur().kind.clone() {
                TokenKind::Import => imports.push(self.parse_import()?),
                TokenKind::Alias => aliases.push(self.parse_alias()?),
                _ => items.push(self.parse_item()?),
            }
        }

        Ok(Program {
            imports,
            aliases,
            items,
        })
    }

    fn parse_import(&mut self) -> Result<Import, String> {
        self.expect(&TokenKind::Import, "Falta 'import'")?;
        let path = match &self.cur().kind {
            TokenKind::StringLit(s) => {
                let p = s.clone();
                self.bump();
                p
            }
            _ => return Err("Se esperaba string en import".to_string()),
        };
        self.expect(&TokenKind::Semicolon, "Falta ';' en import")?;
        Ok(Import { path })
    }

    fn parse_alias(&mut self) -> Result<Alias, String> {
        self.expect(&TokenKind::Alias, "Falta 'alias'")?;
        let name = self.expect_ident("nombre de alias")?;
        self.expect(&TokenKind::Equal, "Falta '=' en alias")?;
        let path = match &self.cur().kind {
            TokenKind::StringLit(s) => {
                let p = s.clone();
                self.bump();
                p
            }
            _ => return Err("Se esperaba string en alias".to_string()),
        };
        self.expect(&TokenKind::Semicolon, "Falta ';' en alias")?;
        Ok(Alias { name, path })
    }

    fn parse_item(&mut self) -> Result<Item, String> {
        let is_extern = self.eat(&TokenKind::Extern);
        match self.cur().kind {
            TokenKind::Struct => Ok(Item::Struct(self.parse_struct(is_extern)?)),
            TokenKind::Fn => Ok(Item::Function(self.parse_function(is_extern)?)),
            _ => Err(format!(
                "Item inválido en {}:{}",
                self.cur().line,
                self.cur().col
            )),
        }
    }

    fn parse_struct(&mut self, is_extern: bool) -> Result<StructDecl, String> {
        self.expect(&TokenKind::Struct, "Falta 'struct'")?;
        let name = self.expect_ident("nombre de struct")?;
        self.expect(&TokenKind::LBrace, "Falta '{' en struct")?;

        let mut fields = Vec::new();
        while self.cur().kind != TokenKind::RBrace {
            let ty = self.parse_type()?;
            let field_name = self.expect_ident("nombre de campo")?;
            self.expect(&TokenKind::Semicolon, "Falta ';' en campo")?;
            fields.push(FieldDecl {
                ty,
                name: field_name,
            });
        }
        self.expect(&TokenKind::RBrace, "Falta '}' en struct")?;
        self.expect(&TokenKind::Semicolon, "Falta ';' tras struct")?;

        Ok(StructDecl {
            name,
            fields,
            is_extern,
        })
    }

    fn parse_function(&mut self, is_extern: bool) -> Result<FunctionDecl, String> {
        self.expect(&TokenKind::Fn, "Falta 'fn'")?;
        let name = self.expect_ident("nombre de función")?;
        self.expect(&TokenKind::LParen, "Falta '('")?;

        let mut params = Vec::new();
        if self.cur().kind != TokenKind::RParen {
            loop {
                let ty = self.parse_type()?;
                let pname = self.expect_ident("nombre de parámetro")?;
                params.push(ParamDecl { ty, name: pname });
                if !self.eat(&TokenKind::Comma) {
                    break;
                }
            }
        }
        self.expect(&TokenKind::RParen, "Falta ')' en función")?;
        self.expect(&TokenKind::Arrow, "Falta '->' en función")?;
        let ret_ty = self.parse_type()?;

        if self.eat(&TokenKind::Semicolon) {
            return Ok(FunctionDecl {
                name,
                params,
                ret_ty,
                body: None,
                is_extern: true,
            });
        }

        self.expect(&TokenKind::LBrace, "Falta '{' en cuerpo")?;
        let mut body = Vec::new();
        while self.cur().kind != TokenKind::RBrace {
            body.push(self.parse_stmt()?);
        }
        self.expect(&TokenKind::RBrace, "Falta '}' en cuerpo")?;

        Ok(FunctionDecl {
            name,
            params,
            ret_ty,
            body: Some(body),
            is_extern,
        })
    }

    fn parse_stmt(&mut self) -> Result<Stmt, String> {
        match self.cur().kind {
            TokenKind::Let => {
                self.bump();
                let ty = self.parse_type()?;
                let name = self.expect_ident("nombre de variable")?;
                self.expect(&TokenKind::Equal, "Falta '=' en let")?;
                let value = self.parse_expr()?;
                self.expect(&TokenKind::Semicolon, "Falta ';' en let")?;
                Ok(Stmt::Let { ty, name, value })
            }
            TokenKind::Return => {
                self.bump();
                let e = self.parse_expr()?;
                self.expect(&TokenKind::Semicolon, "Falta ';' en return")?;
                Ok(Stmt::Return(e))
            }
            _ => {
                let e = self.parse_expr()?;
                self.expect(&TokenKind::Semicolon, "Falta ';' en expresión")?;
                Ok(Stmt::Expr(e))
            }
        }
    }

    fn parse_expr(&mut self) -> Result<Expr, String> {
        self.parse_add_sub()
    }

    fn parse_add_sub(&mut self) -> Result<Expr, String> {
        let mut node = self.parse_mul_div()?;
        loop {
            let op = match self.cur().kind {
                TokenKind::Plus => BinOp::Add,
                TokenKind::Minus => BinOp::Sub,
                _ => break,
            };
            self.bump();
            let rhs = self.parse_mul_div()?;
            node = Expr::Binary {
                op,
                lhs: Box::new(node),
                rhs: Box::new(rhs),
            };
        }
        Ok(node)
    }

    fn parse_mul_div(&mut self) -> Result<Expr, String> {
        let mut node = self.parse_primary()?;
        loop {
            let op = match self.cur().kind {
                TokenKind::Star => BinOp::Mul,
                TokenKind::Slash => BinOp::Div,
                _ => break,
            };
            self.bump();
            let rhs = self.parse_primary()?;
            node = Expr::Binary {
                op,
                lhs: Box::new(node),
                rhs: Box::new(rhs),
            };
        }
        Ok(node)
    }

    fn parse_primary(&mut self) -> Result<Expr, String> {
        match self.cur().kind.clone() {
            TokenKind::IntLit(n) => {
                self.bump();
                Ok(Expr::Int(n))
            }
            TokenKind::Ident(name) => {
                self.bump();
                if self.eat(&TokenKind::LParen) {
                    let mut args = Vec::new();
                    if self.cur().kind != TokenKind::RParen {
                        loop {
                            args.push(self.parse_expr()?);
                            if !self.eat(&TokenKind::Comma) {
                                break;
                            }
                        }
                    }
                    self.expect(&TokenKind::RParen, "Falta ')' en llamada")?;
                    Ok(Expr::Call { name, args })
                } else {
                    Ok(Expr::Var(name))
                }
            }
            TokenKind::LParen => {
                self.bump();
                let expr = self.parse_expr()?;
                self.expect(&TokenKind::RParen, "Falta ')' agrupación")?;
                Ok(expr)
            }
            _ => Err(format!(
                "Expresión inválida en {}:{}",
                self.cur().line,
                self.cur().col
            )),
        }
    }
}
